### PR TITLE
Switch to GOV.UK design system

### PR DESCRIPTION
**Pull request by ushkarev**
MTP apps used to use a combination of [govuk-template](https://github.com/alphagov/govuk_template), [govuk-elements](https://github.com/alphagov/govuk_elements) and [govuk-frontend-toolkit](https://github.com/alphagov/govuk_frontend_toolkit). These are deprecated in favour of the more modern, flexible and accessible [govuk-frontend](https://github.com/alphagov/govuk-frontend).

Switching to the new framework involves changing ~pretty much~ every template, style and javascript module because govuk-frontend uses completely different patterns. It was decided that, during this upgrade, we should minimise our use of customised MTP-specific components where possible in favour of those provided by GDS.

There are still some custom components (used only by [cashbook](https://github.com/ministryofjustice/money-to-prisoners-cashbook) and [noms-ops](https://github.com/ministryofjustice/money-to-prisoners-noms-ops)) that have not been upgraded.

### CODE REVIEW

File: mtp_common/user_admin/views.py [hunk 1]
```diff
@@ -66,14 +66,17 @@ def list_users(request):
...
     user_count = response.get('count', 0)
     users = response.get('results', [])
-    account_requests = retrieve_all_pages_for_path(session, 'requests/')
-    for account_request in account_requests:
-        account_request['created'] = parse_datetime(account_request['created'])
+    try:
+        account_requests = retrieve_all_pages_for_path(session, 'requests/')
+        for account_request in account_requests:
+            account_request['created'] = parse_datetime(account_request['created'])
+    except HttpClientError:
```
Comment (xoen) [inline mtp_common/user_admin/views.py:88]: I'm probably missing something here but worth checking, are we sure `retrieve_all_pages_for_path()` raises a `HttpClientError` here?

`session`, used to make the request is an instance of `auth.api_client.MoJOAuth2Session` (returned by [`api_client.get_api_session()`](https://github.com/ministryofjustice/money-to-prisoners-common/blob/master/mtp_common/user_admin/views.py#L61) on line 63 here). AFAICT the 3rd party library used, makes these requests [using `requests`](https://github.com/requests/requests-oauthlib/blob/master/requests_oauthlib/oauth2_session.py#L521-L523).

I may completely wrong as the classes hierarchy is not that shallow but maybe this would raise a `requests.exceptions.RequestException` of some sort instead? 🤷 

But maybe I'm talking rubbish. I've noticed [this `hooks` thing](https://github.com/ministryofjustice/money-to-prisoners-common/blob/905023f9e99468dd90d984a8ecc1c25deb6c08ec/mtp_common/auth/api_client.py#L73) in `auth.api_client.MoJOAuth2Session` pointing to [`auth.api_client. error_status_response_hook()`](https://github.com/ministryofjustice/money-to-prisoners-common/blob/905023f9e99468dd90d984a8ecc1c25deb6c08ec/mtp_common/auth/api_client.py#L58)  which may well explain why this exception is raised rather than the `responses` one?

Comment (xoen) [reply mtp_common/user_admin/views.py:88]: I'm guessing this specific problem here is for cases when `responses` doesn't raise an exception but the response itself was a `4xx` so this must be right.

I'm probably answering my own question.

**Review by xoen — APPROVED** [review]
I had a look at this, AFAICT LGTM but it's a huge PR and I'm not that familiar with GDS Design System so not sure how much I could really spot here.

File: mtp_common/assets-src/mtp_common/components/cookie-banner/index.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  cookieName: 'seen_cookie_message',
+  selector: '.mtp-cookie-banner',
+
+  init: function () {
+    var $banner = $(this.selector);
+    if ($banner && Cookie.get(this.cookieName) === undefined) {
+      $banner.show();
+    }
+    Cookie.set(this.cookieName, 'yes', {
+      expires: 28
```
Comment (Staberinde) [inline mtp_common/assets-src/mtp_common/components/cookie-banner/index.js:15]: Did we set a cookie expiry previously? Am I right in thinking this will expire session storage for logged in users (which is obv just going to be within HMPPS) every month, an are we okay with that?

File: mtp_common/assets-src/mtp_common/components/hidden-long-text/index.js [hunk 3]
```diff
@@ -1,9 +1,8 @@
-// Hides a long piece of text behind a "more..." link
```
Comment (Staberinde) [inline mtp_common/assets-src/mtp_common/components/hidden-long-text/index.js:1]: No biggy, just wondering if there was a reason for removing this

File: mtp_common/assets-src/mtp_common/components/mailcheck-warning/index.js [hunk 4]
```diff
@@ -1,15 +1,14 @@
 // Show a warning when an email address domain might be misspelled
-/* globals django */
 'use strict';
 
-var mailcheck = require('mailcheck');
+import Mailcheck from 'mailcheck';
```
Comment (Staberinde) [inline mtp_common/assets-src/mtp_common/components/mailcheck-warning/index.js:4]: Are we okay with using es6 syntax here, is this used by the public site?

File: mtp_common/assets-src/mtp_common/components/sub-navigation/_index.scss [hunk 5]
```diff
@@ -0,0 +1,71 @@
+// based on:
+// https://github.com/alphagov/govuk-design-system/blob/5bf961e95ec9cf85f6d607b48f2e2bd9bd29c98b/src/stylesheets/components/_navigation.scss
+// https://github.com/alphagov/govuk-prototype-kit-docs/blob/master/src/stylesheets/components/_navigation.scss
+
+$mtp-sub-navigation-height: 50px;
+$mtp-sub-navigation-current-page-border-size: $govuk-border-width;  // TODO: or $govuk-border-width-narrow
```
Comment (Staberinde) [inline mtp_common/assets-src/mtp_common/components/sub-navigation/_index.scss:6]: Is this TODO still relevant?

**Review by Staberinde — COMMENTED** [review]
168/282 reviewed: Looks good, just a few questions/comments

This might be more of a question for the service-specific PR's, but have you run the functional test suite against these changes for any of the apps yet?

File: mtp_common/assets-src/mtp_common/_display.scss [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  textarea {
+    resize: vertical;
+    min-height: 40px;
+  }
+}
+
+// no govuk-frontend class is available for the inverse of .govuk-!-display-none-print
+.mtp-\!-display-print-only {
```
[the remark this answers is not in the document]
Comment (ushkarev) [reply mtp_common/assets-src/mtp_common/_display.scss:34]: it's how GDS do it e.g. `govuk-!-display-none-print` so i made `mtp-!-display-print-only` to be symmetrical. while it does need to be escaped in selectors in sass files, it's written plainly in html/js.
somewhat unusual but they have loads of these exclamations in various classes so i figure they checked that it works in all browsers.

File: mtp_common/assets-src/mtp_common/components/cookie-banner/index.js (hunk 2 above)
Comment (ushkarev) [reply mtp_common/assets-src/mtp_common/components/cookie-banner/index.js:15]: this is for the banner at the top of staff-only websites that says we use cookies. it used to be part of govuk-template (but no longer exists in govuk-frontend) and i've kept the expiry at the same 28 days. you'd only ever see it a second time if you don't visit the site for _more_ than 28 days:
https://github.com/alphagov/govuk_template/blob/master/source/assets/javascripts/cookie-bar.js#L13

File: mtp_common/assets-src/mtp_common/components/hidden-long-text/index.js (hunk 3 above)
Comment (ushkarev) [reply mtp_common/assets-src/mtp_common/components/hidden-long-text/index.js:1]: the comment? i dunno :)
i went through a phase of trying to add usage notes to the HTML includes and template tags, but they're not totally consistent. it'll be a good idea to tidy them up at some point

File: mtp_common/assets-src/mtp_common/components/mailcheck-warning/index.js (hunk 4 above)
Comment (ushkarev) [reply mtp_common/assets-src/mtp_common/components/mailcheck-warning/index.js:4]: the imports throughout are only there to tell webpack which files need access to which others – it effectively gets edited out when compiled. in the end, you get one gigantic javascript file with no imports.

i've been a bit cautious and not added modern javascript _inside_ what'd be left. that's why it's all still `var` etc rather than `let` etc.

also, cloud platform's setup mean versions of IE below 11 probably won't be able to connect so in theory modern JS might be alright. but i didn't want to go too far!

File: mtp_common/tasks.py [hunk 7]
```diff
@@ -89,7 +89,7 @@ def prepare_email(from_address, to, subject, text_template, html_template, templ
         from_email=from_address,
         to=to
     )
-    if anymail_tags and isinstance(email, AnymailMessage):
+    if anymail_tags and AnymailMessage is not None and isinstance(email, AnymailMessage):
```
[the remark this answers is not in the document]
Comment (ushkarev) [reply mtp_common/tasks.py:92]: you get a TypeError if you do `isinstance(123, None)`. i was just being cautious, i hadn't actually noticed that happening at any point.

File: mtp_common/assets-src/mtp_common/components/sub-navigation/_index.scss (hunk 5 above)
Comment (ushkarev) [reply mtp_common/assets-src/mtp_common/components/sub-navigation/_index.scss:6]: ah so GDS use the thinner border (the size in the comment) but i thought it looked nicer with a thicker one. so far, this component isn't used anywhere (it'll probably first appear in the cashbook) and i was planning on asking people which they thought was better.

File: mtp_common/tasks.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ushkarev) [reply mtp_common/tasks.py:92]: `isinstance('a@example.com', (None,))` doesn't work either. using `NoneType` (i.e. `None.__class__`) _would_ have worked but then the import lines need tweaking

File: mtp_common/assets-src/mtp_common/components/print-trigger/index.js [hunk 8]
```diff
@@ -1,10 +1,9 @@
-// Print module with confirmation dialogue box integration
 'use strict';
 
-var cookie = require('js-cookie');
+import Cookie from 'js-cookie';
```
Comment (Staberinde) [inline mtp_common/assets-src/mtp_common/components/print-trigger/index.js:3]: Same comment about es6 as above

File: tests/urls.py [hunk 9]
```diff
@@ -67,11 +69,17 @@ def __init__(self, *args, **kwargs):
             'template_name': 'mtp_common/auth/reset-password-done.html',
         }, name='reset_password_done'
     ),
+    url(
```
Comment (Staberinde) [inline tests/urls.py:72]: Why is this needed now when it wasn't before?

**Review by Staberinde — COMMENTED** [review]
282/282 reviewed. Nothing especially blocking, just more questions

File: mtp_common/templates/mtp_common/build_tasks/docker-compose.yml [hunk 10]
```diff
@@ -12,4 +12,4 @@ django:
     ENV: "local"
     DEBUG: "True"
     DJANGO_SECRET_KEY: "CHANGE_ME"
-    API_URL: "http://${HOST_MACHINE_IP}:8000"
+    API_URL: "http://127.0.0.1:8000"
```
[the remark this answers is not in the document]
Comment (ushkarev) [reply mtp_common/templates/mtp_common/build_tasks/docker-compose.yml:15]: i did this cos `docker-machine` is no longer a thing. you used to need it when running docker on a mac. i haven't run this build task in years so don't know if it still works anymore. it was there for running all the apps locally in docker (but without any volumes).

we've got the option of completely removing it, but i didn't want to go overboard with these changes.

File: mtp_common/templates/mtp_common/build_tasks/eslintrc.json [hunk 11]
```diff
@@ -14,6 +18,9 @@
     "eol-last": ["error"],
     "max-len": ["error", 120],
     "radix": ["error", "always"],
+    "semi": ["error"],
```
[the remark this answers is not in the document]
Comment (ushkarev) [reply mtp_common/templates/mtp_common/build_tasks/eslintrc.json:21]: no, warnings do not cause tests to fail and as you found these tests don't even get run in CI! that's something we still need to fix.

while it's a change to the config file, i think we all already would have used semicolons in JS – omitting them is very error-prone. i saw this more as making the rules we already use instinctively more explicit.

File: mtp_common/templates/mtp_common/build_tasks/package.json [hunk 12]
```diff
@@ -16,15 +16,15 @@
...
+    "eslint": "~7.17",
+    "sass-lint": "~1.13"
   },
   "dependencies": {
-    "govuk-elements-sass": "~3.1",
-    "jquery": "~1.12",
+    "govuk-frontend": "~3.10",
+    "jquery": "~3.5",
     "js-cookie": "~2.2",
-    "node-sass": "~4.12",
```
[the remark this answers is not in the document]
Comment (ushkarev) [reply mtp_common/templates/mtp_common/build_tasks/package.json:26]: `libsass` as a python dependency: https://github.com/ministryofjustice/money-to-prisoners-common/pull/399/commits/1a19aebe2c5ce61e178ebb4e8d74da30e8ea019d

`node-sass` has very fussy version requirements and was the reason that you couldn't use node versions above 10.

File: tests/urls.py (hunk 9 above)
Comment (ushkarev) [reply tests/urls.py:72]: the settings page didn't used to have a base class in this library – i brought it in here to enable breadcrumbs to work properly. the new tests in https://github.com/ministryofjustice/money-to-prisoners-common/pull/399/commits/86ae6f822a1d76fee591819e67d35c240055f99f now actually load the template (that already existed here) so we need to also include views that it references

File: mtp_common/user_admin/views.py (hunk 1 above)
Comment (ushkarev) [reply mtp_common/user_admin/views.py:88]: yes that's right. that hook wraps the normal `requests` errors with a custom `HttpClientError` type. i made this change because i actually saw the error coming up in the bank-admin site with a user who had some permissions but not others.

File: mtp_common/assets-src/mtp_common/components/print-trigger/index.js (hunk 8 above)
Comment (ushkarev) [reply mtp_common/assets-src/mtp_common/components/print-trigger/index.js:3]: these don't end up appearing in the compiled JS. see this comment…
https://github.com/ministryofjustice/money-to-prisoners-common/pull/399#discussion_r564411710

File: mtp_common/tasks.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Staberinde) [reply mtp_common/tasks.py:92]: That's because `None` isn't a type, it's an instance of 'NoneType`. So the error is actually that you've given a second argument to `isinstance` that's an instance not a type. `isinstance(None, AnymailMessage)` will just return False as other instance would.

[the remark this answers is not in the document]
Comment (ushkarev) [reply mtp_common/tasks.py:92]: yes that's exactly it, `AnymailMessage` can be None so this line could throw a TypeError previously

[the remark this answers is not in the document]
Comment (Staberinde) [reply mtp_common/tasks.py:92]: The type error will only be thrown if this if `None` is the second argument, not the first. Let's have a call to chat about this

[the remark this answers is not in the document]
Comment (ushkarev) [reply mtp_common/tasks.py:92]: that's right, that's why we need the `AnymailMessage is not None` guard for the _second_ argument

[the remark this answers is not in the document]
Comment (Staberinde) [reply mtp_common/tasks.py:92]: I see, there's an `ImportError` guard against `anymail` not being present. Would we expect this code to be executed in the case `Anymail` is not present? I can see why that would be the case in tests, but if that's the case wouldn't we just mock `Anymail` anyway as that's the functionality under test?

[the remark this answers is not in the document]
Comment (ushkarev) [reply mtp_common/tasks.py:92]: `django-anymail` isn't listed in the dependencies of mtp-common so it's possible for this library to be used in a way where `AnymailMessage` is None and email is an instance of `EmailMultiAlternatives` instead.

remember how we were planning on moving _all_ shared dependencies into mtp-common – i'm thinking we can put `django-anymail` into here then. but until then, i simply added this guard to make this line work. cos otherwise you can get a TypeError
